datacube/utils/generic.py

```diff
@@ -26,3 +39,64 @@ def map_with_lookahead(it, if_one=None, if_many=None):
 
     for v in itertools.chain(iter(p1), it):
         yield proc(v)
+
+
+def qmap(func, q, eos_marker=EOS):
+    """ Converts queue to an iterator.
+
+    For every `item` in the `q` that is not `eos_marker`, `yield proc(item)`
+
+    Takes care of calling `.task_done()` on every item extracted from the queue.
+    """
+    while True:
+        item = q.get(block=True)
+        if item is eos_marker:
+            q.task_done()
+            break
+        else:
+            try:
+                yield func(item)
+            finally:
+                q.task_done()
+
+
+def it2q(its, q, eos_marker=EOS):
+    """ Convert iterator into a Queue
+
+        [1, 2, 3] => [1, 2, 3, eos_marker]
+    """
+    try:
+        for x in its:
+            q.put(x, block=True)
+    finally:
+        q.put(eos_marker, block=True)
+
+
+def thread_local_cache(name: str,
```

Comment:
I'm finding this function quite confusing. I think it would be a lot clearer if _purge_ was a different function. Also, consider _pop_ as a name, [as in dicts](https://docs.python.org/3/library/stdtypes.html#dict.pop).

Reply:
`purge` option is really just to enable unit testing, I don't like multi-function solution, might as well make a class/singleton object if trying to model dict.